File: duplicates.c
/*
Exercício feito no whiteboard com os
recrutadores do mercado livre

Foi soilicitado para verificar números duplicados
em um array de ints

Nessa implementação feita em C
provavelmente daria um timeout
devido ao while dentro do while,
sendo necessário percorrer o array
len(array) vezes
(com percurso menor a cada vez)
*/

#include "stdio.h"
#include "stdbool.h"
#include <stdlib.h>

bool	verify_duplicates(int *array, int len)
{
	int	i;
	int	j;

	i = 0;
	while (array[i] != '\0')
	{
		j = i + 1;
		while (array[i] != array[j] && array[j])
			j++;
		if (j < len)
			return (true);
		i++;
	}
	return (false);
}

int	*do_atoi(int argc, char **argv)
{
	int	i;
	int	*array;

	array = malloc(sizeof(int) * argc);
	i = 0;
	while (i < argc - 1)
	{
		array[i] = atoi(argv[i + 1]);
		i++;
	}
	array[i] = '\0';
	return (array);
}

int	main(int argc, char **argv)
{
	int	*array;

	array = do_atoi(argc, argv);
	printf("%d\n", verify_duplicates(array, argc - 1));
	free(array);
}

File: mercado_livre/duplicates.py
"""
Exercício feito no whiteboard com os
recrutadores do mercado livre

Foi soilicitado para verificar números duplicados
em um array de ints

Nessa implementação feita em Python
provavelmente não daria um timeout
já que foi utilizado um dicionário,
sendo necessário percorrer o array
uma única vez
"""

from collections import defaultdict


def duplicate(array):
    x = defaultdict(int)
    for i in range(len(array)):
        x[array[i]] += 1
        if x[array[i]] > 1:
            print(x)
            return True
    print(x)
    return False


if __name__ == '__main__':
    array = [1, 5, 3, 6, 2, 7, 1]
    print(duplicate(array))
